src/mame/misc/babysuprem.cpp

```diff
@@ -33,45 +46,364 @@ class bsuprem_state : public driver_device
 		: driver_device(mconfig, type, tag)
 		, m_maincpu(*this, "maincpu")
 		, m_psg(*this, "psg%u", 1U)
+		, m_nvram(*this, "nvram")
+		, m_dsw1_switches2(*this, "DSW1_SWITCHES2")
+		, m_dsw2(*this, "DSW2")
+		, m_u13_switches(*this, "U13_SWITCHES")
+		, m_test(*this, "TEST")
+		, m_digits(*this, "digit%u", 0U)
+		, m_lamps(*this, "lamp%u", 0U)
+		, m_buttons(*this, "button%u", 0U)
+		, m_nixie(*this, "nixie%u", 0U)
+		, m_hopper_5(*this, "hopper_5")
+		, m_hopper_25(*this, "hopper_25")
+		, m_hopper_100(*this, "hopper_100")
 	{
 	}
 
 	void bsuprem(machine_config &config);
+	void bsuprem_map(address_map &map);
+
+	DECLARE_INPUT_CHANGED_MEMBER(test_pressed);
 
 private:
 	required_device<cpu_device> m_maincpu;
 	required_device_array<ay8910_device, 2> m_psg;
+	required_device<nvram_device> m_nvram;
+	required_ioport m_dsw1_switches2;
+	required_ioport m_dsw2;
+	required_ioport m_u13_switches;
+	required_ioport m_test;
+	output_finder<6> m_digits;
+	output_finder<45> m_lamps;
+	output_finder<5> m_buttons;
+	output_finder<4> m_nixie;
+	required_device<ticket_dispenser_device> m_hopper_5;
+	required_device<ticket_dispenser_device> m_hopper_25;
+	required_device<ticket_dispenser_device> m_hopper_100;
+
+	uint8_t m_irq_state;
+	uint8_t m_u1_porta_data;
+	uint8_t m_u10_data;
+
+	std::unique_ptr<uint8_t[]> m_nvram8;
+
+	uint8_t nvram_r(offs_t offset);
+	void nvram_w(offs_t offset, uint8_t data);
+
+	void ay8910_u1_porta_w(uint8_t data);
+	void ay8910_u1_portb_w(uint8_t data);
+	void ay8910_u2_porta_w(uint8_t data);
+	void ay8910_u2_portb_w(uint8_t data);
+	uint8_t ay8910_w_dsw1_switches2_r(offs_t offset);
+	void ay8910_w(offs_t offset, uint8_t data);
+
+	uint8_t dsw2_u13_switches_r(offs_t offset);
+	void u9_w(offs_t offset, uint8_t data);
+	void u10_w(offs_t offset, uint8_t data);
+
+	virtual void machine_reset() override;
+	virtual void machine_start() override;
+
+	INTERRUPT_GEN_MEMBER(mains_irq);
+
 };
 
 static INPUT_PORTS_START( bsuprem )
-	PORT_START("DSW1")
-	PORT_DIPUNKNOWN_DIPLOC(0x01, 0x01, "SW1:1")
-	PORT_DIPUNKNOWN_DIPLOC(0x02, 0x02, "SW1:2")
-	PORT_DIPUNKNOWN_DIPLOC(0x04, 0x04, "SW1:3")
-	PORT_DIPUNKNOWN_DIPLOC(0x08, 0x08, "SW1:4")
+	PORT_START("DSW1_SWITCHES2")
+	PORT_BIT(0x01, IP_ACTIVE_LOW, IPT_BUTTON3) PORT_NAME("Right stop")
+	PORT_BIT(0x02, IP_ACTIVE_LOW, IPT_BUTTON2) PORT_NAME("Middle stop")
+	PORT_BIT(0x04, IP_ACTIVE_LOW, IPT_BUTTON1) PORT_NAME("Left stop")
+	PORT_BIT(0x08, IP_ACTIVE_LOW, IPT_BUTTON4)
+	PORT_DIPNAME(0x80, 0x80, "SW1-3 (Test 4)")
+	PORT_DIPSETTING(   0x00, DEF_STR( On ))
+	PORT_DIPSETTING(   0x80, DEF_STR( Off ))
+	PORT_DIPNAME(0x40, 0x40, "SW1-1 (Test 2)")
+	PORT_DIPSETTING(   0x00, DEF_STR( On ))
+	PORT_DIPSETTING(   0x40, DEF_STR( Off ))
+	PORT_DIPNAME(0x20, 0x20, "SW1-2 (Test 3)")
+	PORT_DIPSETTING(   0x00, DEF_STR( On ))
+	PORT_DIPSETTING(   0x20, DEF_STR( Off ))
+	PORT_DIPNAME(0x10, 0x10, "SW1-4 (Test 5)")
+	PORT_DIPSETTING(   0x00, DEF_STR( On ))
+	PORT_DIPSETTING(   0x10, DEF_STR( Off ))
```

Comment:
If these are DIP switches, please use DIP location markup properly so switches will be shown on the menu.